src/ProtocolStreamReader.php

```diff
@@ -54,6 +54,9 @@ public function __construct($input)
                             $this->parsingMode = ParsingMode::HEADERS;
                             $this->headers = [];
                             $this->buffer = '';
+
+                            // after reading a full message, leave to allow different tasks to run
+                            return;
```

Comment:
I think I wrote a lot of confusing stuff... :) please disregard everything I said earlier and read my comment about the current state of the event loop.